graph.py

```diff
@@ -0,0 +1,71 @@
+from itertools import product
+import random
+from collections import deque
```

Comment:
I prefer to import the namespace like `import collections` and then we we go to use something do so like `collections.deque`. This reinforces the full paths to things, and helps remember where they are and what they are. There's a _slight_ performance hit if you instantiate _lots_ of things like this, but in general the readability outweighs the hit imho